### ISSUE TITLE
Possible race condition in calculation of cluster tasks duration

### ISSUE THREAD

**Issue by Selutario** [wazuh/wazuh#10038]
|Wazuh version|Component|
|---|---|
| 4.2.0+ | Wazuh cluster |

## Description
I have been analyzing the logs of a couple of large clusters and, after that, I discovered something strange:
```
2021/09/08 10:56:22 INFO: [Worker Test_cluster_performance_B252_manager_1] [Agent-info sync] Starting.
2021/09/08 10:56:22 DEBUG: [Worker Test_cluster_performance_B252_manager_1] [Agent-info sync] Permission to synchronize granted.
2021/09/08 10:56:23 DEBUG: [Worker Test_cluster_performance_B252_manager_1] [Agent-info sync] Obtained 68 chunks of data in 0.207s.
2021/09/08 10:56:31 INFO: [Worker Test_cluster_performance_B252_manager_1] [Agent-info sync] Finished in 22.882s (68 chunks updated).
2021/09/08 10:56:31 DEBUG: [Worker Test_cluster_performance_B252_manager_1] [Agent-info sync] All chunks sent.
```

We can see `[Agent-info sync] Finished in 22.882s`. However, if we look at the timestamp of each log message, it can be seen that the task started at `10:56:22` and ended at `10:56:31`. That is, only 9 seconds have passed, which does not match the 22.882s that were shown in the _Finished in_ log.

I have found this behavior not only in the Agent-info task but also in the Integrity check:
```
2021/09/08 11:02:38 INFO: [Worker Test_cluster_performance_B252_manager_8] [Integrity check] Starting.
2021/09/08 11:02:38 DEBUG: [Worker Test_cluster_performance_B252_manager_8] [Integrity check] Permission to synchronize granted.
2021/09/08 11:02:38 DEBUG: [Worker Test_cluster_performance_B252_manager_8] [Integrity check] Compressing 'files_metadata.json'.
2021/09/08 11:02:38 DEBUG: [Worker Test_cluster_performance_B252_manager_8] [Integrity check] Sending zip file to master.
2021/09/08 11:02:39 DEBUG: [Worker Test_cluster_performance_B252_manager_8] [Integrity check] Zip file sent to master.
2021/09/08 11:02:39 INFO: [Worker Test_cluster_performance_B252_manager_8] [Integrity check] Finished in 12.819s. Sync not required.
```

It is necessary to investigate, detail, and fix the origin of the problem.

**Comment by Selutario** [comment]
## Update

As mentioned in the title, the reason behind this wrong behavior seems to be a race condition. The process followed during synchronization (for example, from agent-info) is as explained below.

### Agent-info workflow
This loop repeats every 10 seconds. Inside is the method in charge of sending agent information to the master (`agent_info.sync()`).
https://github.com/wazuh/wazuh/blob/b1c7ae23cba517c05b64063e8d54121b9e003191/framework/wazuh/core/cluster/worker.py#L606-L619

Said method does not wait for the master to answer whether it has updated its information or not. As soon as the data has been sent, the `self.agent_info_sync_status['date_start']` variable is updated using the timestamp when the task started. Then it waits (10 seconds) until the next iteration takes place.

As `agent_info.sync()` does not wait for the master to respond, it could happen that after 10 seconds, the worker tries to send the information again but the master has not yet completed the previous task. In this case, the master will not accept the synchronization and the `synced` variable (on line #L611) will be `None`. That is the reason why it is updated after running `agent_info.sync()`, because this way it won't crush the timestamp of a previous valid sync. 

The place where it is calculated how long the Agent info task has taken is in the following method, which is executed every time the worker receives the `syn_m_a_e` command from the master (which means it has updated its information):
https://github.com/wazuh/wazuh/blob/b1c7ae23cba517c05b64063e8d54121b9e003191/framework/wazuh/core/cluster/worker.py#L529-L536

As you can see, the timestamp that was saved on `self.agent_info_sync_status['date_start']` (line #L612) is subtracted from the current one (time.time()).

### Race condition
It could be the case where the master responds (`syn_m_a_e`) almost instantaneously as soon as the worker has finished sending chunks:

In progress

**Comment by Selutario** [comment]
## Issue update

We have been coming up with some solutions to this problem, and for the moment (pending the approval of the team), we believe that the following may be the most appropriate.

Since we cannot trust an instance variable in the worker since it can be updated after receiving the response from the master, an alternative would be to send information related to the task itself to the master. We can send the timestamp in which each task starts together with its information (this would be done in `agent-info sync`, `integrity check` and `integrity sync` in case of containing extra-valid files).

For example, in the `agent-info sync` task, we would send the timestamp in which said task started (on the worker) along with `task_id` that notifies the master where the string with the information has been saved. It would look like this:
```PYTHON
            # Specify under which task_id the JSON can be found in the master.
            await self.worker.send_request(command=self.cmd, data=task_id + b' ' + start_time.encode())
            self.logger.debug(f"All chunks sent.")
```

Then, the only thing the master would do with said timestamp would be to send it back to the worker through the `syn_m_a_e` [command](https://documentation.wazuh.com/current/development/wazuh-cluster.html#wazuh-cluster-protocol), which indicates that the synchronization has finished. The worker would use the information that comes from the master (`data['worker_start_time']`) instead of his own to calculate what the real duration of the task is:
```PYTHON
    def sync_agent_info_from_master(self, response) -> Tuple[bytes, bytes]:
        logger = self.task_loggers['Agent-info sync']
        data = json.loads(response)
        msg = f"Finished in {(time.time()-data['worker_start_time']):.3f}s ({data['updated_chunks']} " \
              f"chunks updated)."
        ...
```

Regards.

**Comment by Selutario** [comment]
## Issue update

After discussing it with the team, we have decided that the master should not receive any timestamps or IDs from worker tasks. This change would be difficult to scale in the long term and is not really necessary, since we can have complete control in the worker part of what tasks are in progress and what their duration is.

Although a solution only on the worker side is feasible and relatively straightforward for `Agent-info` and `Integrity check` tasks, the same is not true for `Integrity sync`. Integrity sync starts on the worker every time it receives files from the master. However, if there are also extra-valid files (files that the worker has to send to the master), a new task is created to send those files (#L708):
https://github.com/wazuh/wazuh/blob/b1c7ae23cba517c05b64063e8d54121b9e003191/framework/wazuh/core/cluster/worker.py#L697-L710

The creation and completion of this task is independent of the original one, which is released and therefore, the worker could start a new `integrity check` **->** `integrity sync` process and receive files again, which would overwrite the integrity sync start date. 
https://github.com/wazuh/wazuh/blob/b1c7ae23cba517c05b64063e8d54121b9e003191/framework/wazuh/core/cluster/worker.py#L682

If that happens and the extra-valid process (the task started at #L708) is not finished yet, the task duration will be incorrect:
https://github.com/wazuh/wazuh/blob/b1c7ae23cba517c05b64063e8d54121b9e003191/framework/wazuh/core/cluster/worker.py#L645

## Conclusion
We have decided that the current behavior of integrity sync when there are extra-valid files is not the most suitable, and the following issue has been opened to address it: https://github.com/wazuh/wazuh/issues/10123

In short, it is intended to "integrate" the synchronization of `extra-valid` files within the rest of the files (`missing`, `shared` and `extra`). Therefore, the worker should not be given permission to perform a new integrity check until the extra-valid files, if any, have been sent.

**Comment by Selutario** [comment]
## Issue update

This problem and its solution have been addressed here: [Refactor of the extra-valid-files cluster's task #10123
](https://github.com/wazuh/wazuh/issues/10123)

This issue can be closed.
